---
title: "Data Science for Bioinformatics"
output: html_document
editor_options: 
  chunk_output_type: console
---

#!begin

This code block removes all the answers and saves this file without the ".answers" part of the filename  

So basically when you knit this markdown file it will:  

1. remove all the answers and save that as a file for the students.  
2. Render that faile, so you can check that it looks ok (html)
3. Render this file including all the answers, so you can check that it looks ok (html)

```{r echo=F, results="hide"}
source("../remove_answer_lines_from_exercises.R")
```
#!end

# Drosophila melanogaster Genetic Reference Panel

Reference paper describes the data obtained here. 

Paper source: The Drosophila melanogaster Genetic Reference Panel, Nature 2012

Data source  

 * http://dgrp2.gnets.ncsu.edu/  (the project)
 * http://dgrp2.gnets.ncsu.edu/data.html  (the data source tabular 3)
 
Tabular formatted genotype (space delimited, 0 = ref allele, 2 = alt allele (not necessarily minor), - = missing)

A zipped version of the data is available in the datasets folder.

Unzipped it is about 2 gigabytes and will probably be too large for your computers memory.

But for speed we recommend that you unzip the data - then work on the unzipped data file!

```{r}

unzip(zipfile = "../datasets/dataset.02.dgrp2.tgeno.zip", overwrite = T)
dir()

```


Luckily, you have an idea on how to handle large data...

#### Q: Read the first 2 rows and inspect them 

Hint: ?read_lines()

Hint: Are these data tab, comma or space separated?

Hint: Do the data have column headers?


```{r}

library(tidyverse)
?read_lines()

```


#!begin

```{r}

read_lines(file = "dgrp2.tgeno", n_max = 2, progress = TRUE)

# Estimate number of lines based on size of first 1000 lines
size1000  <- sum(nchar(readLines(con = "dgrp2.tgeno", n = 1010)[10:1010]))
sizetotal <- file.size("dgrp2.tgeno")
1000 *  sizetotal / size1000

```

#!end


#### Q: Now read the first 10 rows in correct format (214 variables)

#!begin

```{r}

read_delim(file = "dgrp2.tgeno",
           n_max = 10, delim = " ", comment = "#", 
           progress = F)

```

#!end


#### Q: Describe briefly the structure of the data and what the different variables mean


# Tidy data

One dataset, the tidy dataset, will be much easier to work with inside the tidyverse.

There are three interrelated rules which make a dataset tidy:

 * Each variable must have its own column.
 * Each observation must have its own row.
 * Each value must have its own cell.

#### Q: Discuss how this dataset should be organized to be in "tidy" format.

When you all agree then try and make a tidy dataset for the first 10 variants only.

Hint: genotype is a single variable, there is one genotype pr. variants pr. line

Hint: ?gather()

Hint: ?tidyselect::select_helpers

#### Q: Make a function that takes your imported variants and make them tidy (call it "make_tidy")

Final hint: You should get 205 rows pr. variant ;)

#!begin

```{r}

library(tidyverse)

df <- read_delim(file = "dgrp2.tgeno", guess_max = 1000, col_names = T, n_max = 10, delim = " ", comment = "#", progress = F)

make_tidy <- function(df) {
  tt <- df %>% gather(key = "line", value = "genotype", starts_with("line"))
  return(tt)
}

tidy_data <- make_tidy(df)

print(tidy_data)

```

#!end

#### Example of why tidy data is so nice: Calculate the amount of missing data for these two first variants

Hopefully you will realize this is easy, once the data are in tidy format.

Then you can do stuff like this:


```{r eval=FALSE}

tidy_data %>%
  group_by(id,genotype) %>%
  summarise(count = n()) %>%
  spread(key=genotype, value=count) %>%
  {.}

```

# Testing genotyping succes

I have two hypotheses:

# 1. All chromosomes have the same distribution of genotypes

Here you should test the NULL hypothesis that the three genotype, ref, alt or missing is independent of which chromosome we are looking at.

HINT: ?stringr::str_sub

# 2. Genotyping success is dependent on coverage (cov), low coverage have more problems in genotyping.

Here we hypothesize that low coverage variants have more NA's than high coverage variants.

I suggest that you bin the coverage in bins of size 0-4, 5-9, 10-14, >=15

Hint: 5 * (0:35 %/% 5)

#### Q: Show, test and visualize the contingency table of genotype (0 or 2) X chr.

#### Q: Show, test and visualize the contingency table of genotype (NA or not NA) X coverage bin.

These are not easy questions!

Hint: you should UNZIP the data, otherwise R will use a lot of memory for reading and unzipping in memory.

#### Example 9.4 to the rescue!

For testing, use chisq.test() - and make sure you do it correctly. First you should redo example 9.4 and check that you can get the same results.

Here is some help:

```{r example_9.4 }

x1 <- data.frame("eaten"    = rep("Eaten by birds", 3),
                "infection" = c("Uninfected", "Lightly infected", "Highly infected"),
                "count"     = c(1,10,37))

x2 <- data.frame("eaten"    = rep("Not eaten by birds", 3),
                "infection" = c("Uninfected", "Lightly infected", "Highly infected"),
                "count"     = c(49,35,9))

x <- rbind(x1,x2)
rm(x1,x2)

xt <- x %>% 
  spread(key=infection, value=count, fill=0) %>%
  data.frame()

rownames(xt) <- xt$eaten
xt <- xt %>% select(-eaten) %>% as.matrix()

knitr::kable(addmargins(xt))

```


```{r}

chisq.test(x=xt, correct = F)

```


```{r}

worm <- read_csv(file="http://www.zoology.ubc.ca/~schluter/WhitlockSchluter/wp-content/data/chapter09/chap09e4WormGetsBird.csv")

print(worm)
chisq.test(x = worm$infection, y=worm$fate, correct = F)

```

```{r}

knitr::kable(addmargins(table(worm$infection, worm$fate)))
chisq.test(x = table(worm$infection, worm$fate), correct=F)

```

You should also visualize the contingency tables, something like figure 9.4-1

HINT: geom_mosaic() from the package ggmosaic - or http://whitlockschluter.zoology.ubc.ca/r-code/rcode09

For both questions, you should be ready to present your approach, results and conclusions.

#!begin

# Example 9.4

```{r}

library(tidyverse)
library(ggmosaic)

ggplot(data = x ) + 
  geom_mosaic(mapping = aes(weight=count, x=product(infection), fill=eaten)) +
  xlab("Infection level") +
  ylab("Relative frequency") + 
  NULL

```

# Reading and tidying data

```{r read and tidy data}

library(tidyverse)

f <- function(df, pos) {
  tt <- df %>% 
    mutate(covbin = 5* (cov %/% 5),                                    # make bins of coverage
           chromosome = str_sub(chr, start=1, end=1)) %>%              # extract chromosome
    gather(key = "line", value = "genotype", starts_with("line")) %>%  # make tidy format
    group_by(chromosome, genotype, covbin) %>%
    summarise(count = n()) %>%
    {.}
  
  cat("Position", pos, "Time:", as.character(Sys.time()), "\n", file = "test.log", append = TRUE) # Log progress to log file
  
  return(tt)
}

df <- read_delim_chunked(file = "dgrp2.tgeno", 
                         callback = ListCallback$new(f), 
                         chunk_size = 50000, guess_max = 100,
                         col_names = T, delim = " ", comment = "#",
                         progress = F)

df <- bind_rows(df)

```

# Chromosome and ref/alt genotype counts

Here we summarize data for each chromosome arm and genotype

```{r chromosome and genotype counts}

x <- df %>%
  ungroup() %>%
  filter(genotype %in% c("0","2", "-")) %>%
  group_by(chromosome, genotype) %>%
  summarise(count = sum(count)) %>%
  ungroup() %>%
  {.}

xt <- x %>% 
  spread(key=genotype, value=count, fill=0) %>%
  as.data.frame()

rownames(xt) <- xt$chromosome
xt <- xt %>% 
  select(-chromosome) %>% 
  as.matrix()

knitr::kable(addmargins(xt))

chisq.test(x=xt, correct=F)

ggplot(data = x ) + 
  geom_mosaic(mapping = aes(weight=count, x=product(chromosome), fill=genotype)) +
  xlab("Chromosome") +
  ylab("Relative frequency") + 
  NULL

```

# Coverage and quality

Here we summarize data for each coverage bin and genotype sucess/failure

```{r coverage and quality}

x <- df %>%
  ungroup() %>%
  mutate(genotype = ifelse(is.na(genotype), "NA", "Not NA"),
         covbin   = ifelse(covbin >=15, 15, covbin)) %>%
  group_by(covbin, genotype) %>%
  summarise(count = sum(count, na.rm=T)) %>%
  ungroup() %>%
  {.}

xt <- x %>% 
  spread(key=genotype, value=count, fill=0) %>%
  as.data.frame()

rownames(xt) <- xt$covbin
xt <- xt %>% 
  select(-covbin) %>% 
  as.matrix()

knitr::kable(addmargins(xt))

chisq.test(x=xt, correct=F)

ggplot(data = x ) + 
  geom_mosaic(mapping = aes(weight=count, x=product(covbin), fill=genotype)) +
  xlab("Coverage bin") +
  ylab("Relative frequency") + 
  NULL


```

#!end

# Getting my disk space back

```{r}

# Delete the unzipped file when everything is done
unlink("dgrp2.tgeno", force = TRUE)

```

